packages/cozy-scanner/src/Scanner.jsx

```diff
@@ -0,0 +1,190 @@
+import React from 'react'
+import { withClient } from 'cozy-client'
+import PropTypes from 'prop-types'
+import { CozyFile } from 'cozy-doctypes'
+import { Modal } from 'cozy-ui/transpiled/react'
+
+import ScannerQualification from './ScannerQualification'
+import withOffline from 'cozy-ui/transpiled/helpers/withOffline'
+export const SCANNER_IDLE = 'idle'
+export const SCANNER_DONE = 'done'
+export const SCANNER_UPLOADING = 'uploading'
+/**
+ * LoadingScreen is used to create a better transition between
+ * the native camera and the Qualification Modale.
+ *
+ * This Overlay has to be in a Modal since the menu item can be
+ * closed
+ *
+ */
+class Scanner extends React.Component {
+  state = {
+    status: SCANNER_IDLE,
+    error: null,
+    filename: '',
+    shouldShowScannerQualification: false,
+    imageURI: '',
+    loadingScreen: false
+  }
+  constructor(props) {
+    super(props)
+    if (!CozyFile.cozyClient) CozyFile.registerClient(this.props.client)
+  }
+
+  /**
+   *
+   * @param {String} imageURI native path to the file (file:///var....)
+   */
+  onSuccess = async imageURI => {
+    this.setState({
+      error: null,
+      loadingScreen: false,
+      shouldShowScannerQualification: true,
+      imageURI
+    })
+  }
+
+  onUpload = async (imageURI, qualification, filename = '') => {
+    const { generateName } = this.props
+    const name = filename === '' ? generateName() : filename
+    this.setState({ status: SCANNER_UPLOADING, filename: name })
+    const { dirId, onConflict, onBeforeUpload, onFinish } = this.props
+    const onResolvedLocalFS = async fileEntry => {
+      fileEntry.file(
+        file => {
+          const reader = new FileReader()
+          reader.onloadend = async () => {
+            //we get the result of the readAsBuffer in the `result` attr
+            try {
+              if (onBeforeUpload) onBeforeUpload()
+              const newFile = await CozyFile.uploadFileWithConflictStrategy(
+                name,
+                reader.result,
+                dirId,
+                onConflict,
+                qualification
+              )
+              //It's possible that the filename as changed, so let's update it
+              if (onConflict === 'rename') {
+                this.setState({ filename: newFile.data.name })
+              }
+
+              if (onFinish) onFinish()
+            } catch (error) {
+              this.setState({ error })
+            } finally {
+              this.setState({ status: SCANNER_DONE })
+            }
+          }
+          // Read the file as an ArrayBuffer
+          reader.readAsArrayBuffer(file)
+        },
+        err => {
+          this.setState({ error: err })
+          this.setState({ status: SCANNER_DONE })
+          //Since this module is pretty recent, let's have this info in sentry if needed
+          console.error('error getting fileentry file!' + err) //eslint-disable-line no-console
+        }
+      )
```

Comment:
This function seems pretty complex and is not tested as far as I am aware. It seems to deal with several things.

- state of the component
- reading the file
- uploading the file

Do you think we could we extract the file reading and uploading logic out the component so that the component only deal with its state and props ?

```
     this.setState({ status: SCANNER_UPLOADING, filename: name })
     try {
     ... do before upload stuff ...
      const { data: newFile } = await doUpload(imageURI)
      this.setState({ filename: newFile.name }) 
     } catch (e) {
       this.setState({ status: SCANNER_DONE, error: e }) 
     } finally {
       this.setState({ status: SCANNER_DONE }) 
      }
```